client/meter/blockchain.js

```diff
@@ -7,6 +7,78 @@ const bigNumber = require('bignumber.js');
 
 const EnergyMaster = bonds.makeContract(ENERGY_MASTER_ADDRESS, ENERGY_MASTER_ABI);
 
+const UPDATE_INTERVAL = 1000;
+
+let contracts = {};
+let buyerContractsSet = new Set();
+let sellerContractsSet = new Set();
+let availableContractsSet = new Set();
+let inited = false;
+let lastBlock = 0;
+let lastCount = 0;
+
+async function getContractInfoByAddress(address) {
+  const account = await myAccount();
+  const contract = makeEnergyContract(address);
+  const [
+    seller,
+    offeredAmount,
+    unitPrice,
+    remainingAmount
+  ] = await Promise.all([
+    contract.seller(),
+    contract.offeredAmount(),
+    contract.unitPrice(),
+    contract.remainingEnergy(account)
+  ]);
+  contracts[address] = {
+    seller,
+    offeredAmount,
+    unitPrice,
+    remainingAmount
+  };
+  if (seller === account && offeredAmount > 0) {
+    sellerContractsSet.add(address);
+  } else {
+    sellerContractsSet.delete(address);
+  }
+  if (remainingAmount > 0) {
+    buyerContractsSet.add(address);
+  } else {
+    buyerContractsSet.delete(address);
+  }
+  if (offeredAmount > 0) {
+    availableContractsSet.add(address);
+  } else {
+    availableContractsSet.delete(address);
+  }
+}
+
+async function getContractInfoByIndex(index) {
+  const contractEntity = await EnergyMaster.contracts(index);
+  const contractAddr = contractEntity[0];
+  await getContractInfoByAddress(contractAddr);
+}
+
+async function updateMaster() {
+  const count = await EnergyMaster.contractCount();
+  const promises = [];
+  for (let i = lastCount; i < count; i++) {
+    promises.push(getContractInfoByIndex(i));
+  }
+  await Promise.all(promises);
+  lastCount = count;
+}
+
+async function init() {
+  // FIXME: Multiple inits
```

Comment:
Explanation: Due to javascript event loop, `init` can get called multiple times due to `inited` not being true. This should be benign but is confusing.

Note: set `inited = true` at the beginning will not work, because it will unblock further operations and unsynced data will be retrieved.

Reply:
Leaving it for now, will create new issue.

Reply:
js is cancer